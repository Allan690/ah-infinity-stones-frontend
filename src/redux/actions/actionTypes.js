export const REQUEST_LOGIN = 'REQUEST_LOGIN';
export const RECEIVE_LOGIN = 'RECEIVE_LOGIN';
export const LOGIN_ERROR = 'LOGIN_ERROR';
export const SIGNUP_REQUEST = 'SIGNUP_REQUEST';
export const SIGNUP_SUCCESS = 'SIGNUP_SUCCESS';
export const SIGNUP_FAILURE = 'SIGNUP_FAILURE';

export const SOCIAL_LOGIN_REQUEST = 'SOCIAL_LOGIN_REQUEST';
export const SOCIAL_LOGIN_SUCCESS = 'SOCIAL_LOGIN_SUCCESS';
export const SOCIAL_LOGIN_ERROR = 'SOCIAL_LOGIN_ERROR';


export const REQUEST_INCREMENT = 'REQUEST_INCREMENT';
export const RECEIVE_INCREMENT = 'RECEIVE_INCREMENT';

export const PASSWORD_RESET_START = 'PASSWORD_RESET_START';
export const PASSWORD_RESET_COMPLETE = 'PASSWORD_RESET_COMPLETE';
export const PASSWORD_RESET_ERROR = 'PASSWORD_RESET_ERROR';


export const PASSWORD_RESET_UPDATE_START = 'PASSWORD_RESET_UPDATE_START';
export const PASSWORD_RESET_UPDATE_COMPLETE = 'PASSWORD_RESET_UPDATE_COMPLETE';
export const PASSWORD_RESET_UPDATE_ERROR = 'ASSWORD_RESET_UPDATE_ERROR';
