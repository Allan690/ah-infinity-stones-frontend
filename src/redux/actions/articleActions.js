import {
  CREATE_ARTICLE, CREATE_ARTICLE_SUCCESS, CREATE_ARTICLE_FAILURE,
  GET_ALL_ARTICLES_REQUEST, GET_ALL_ARTICLES_SUCCESS, GET_ALL_ARTICLES_FAILURE,
  GET_ONE_ARTICLE_REQUEST, GET_ONE_ARTICLE_SUCCESS, GET_ONE_ARTICLE_FAILURE,
  EDIT_ARTICLE, UPDATE_ARTICLE, DELETE_ARTICLE,
} from './actionTypes';


export const createArticleAction = article => ({
  type: CREATE_ARTICLE,
  payload: article,
});
export const createArticleSuccessAction = article => ({
  type: CREATE_ARTICLE_SUCCESS,
  payload: article,
});
export const createArticleFailureAction = error => ({
  type: CREATE_ARTICLE_FAILURE,
  payload: error,
});

export const getAllArticlesRequestAction = () => ({
  type: GET_ALL_ARTICLES_REQUEST,
});

export const getAllArticlesSuccessAction = articles => ({
  type: GET_ALL_ARTICLES_SUCCESS,
  payload: [articles],
});

export const getAllArticlesFailureAction = error => ({
  type: GET_ALL_ARTICLES_FAILURE,
  payload: error,
});

export const getOneArticleRequestAction = artSlug => ({
  type: GET_ONE_ARTICLE_REQUEST,
  payload: artSlug,
});

export const getOneArticleSuccessAction = article => ({
  type: GET_ONE_ARTICLE_SUCCESS,
  payload: article,
});

export const getOneArticleFailureAction = error => ({
  type: GET_ONE_ARTICLE_FAILURE,
  payload: error,
});

export const editArticleAction = article => ({
  type: EDIT_ARTICLE,
  payload: article,
});

export const updateArticleAction = article => ({
  type: UPDATE_ARTICLE,
  payload: article,
});

export const deleteArticleAction = art_slug => ({
  type: DELETE_ARTICLE,
  payload: art_slug,
});
