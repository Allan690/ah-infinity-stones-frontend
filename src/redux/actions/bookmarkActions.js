import {
  BOOKMARK_REQUEST,
  BOOKMARK_SUCCESS,
  BOOKMARK_ERROR,
} from './actionTypes';

export const bookmarkRequest = slug => ({
  type: BOOKMARK_REQUEST,
  slug,
});

export const bookmarkSuccess = data => ({
  type: BOOKMARK_SUCCESS,
  data,
});

export const bookmarkError = error => ({
  type: BOOKMARK_ERROR,
  error,
});
