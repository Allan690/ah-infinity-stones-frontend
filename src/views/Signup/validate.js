/* eslint-disable no-useless-escape */
/**
 *
 * @param {*} data
 * @returns msg
 */
export default function validateData(data) {
  const emailPattern = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
  const strongPassword = new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])(?=.{8,})');
  const msg = {
    usernameError: '',
    emailError: '',
    passwordError: '',
    cpasswordError: '',
    successMsg: '',
  };
  if (data.username.length > 0 && data.username.length < 6) {
    msg.usernameError = 'Username must have at least six characters';
  } else if (data.email.length > 0 && !emailPattern.test(data.email)) {
    msg.emailError = 'Invalid email address';
  } else if (data.password.length > 0 && data.password.length < 6) {
    msg.passwordError = 'Password must have at least six characters';
  } else if (data.password.length > 0 && !strongPassword.test(data.password)) {
    msg.passwordError = 'Enter strong password e.g p!Yd7YzU,S8}p}@';
  } else if (data.cpassword.length > 0 && data.cpassword !== data.password) {
    msg.cpasswordError = 'Password does not match';
  }
  return msg;
}

/**
 *
 * @param {*} data
 * @returns buttonDisabled
 */
export function handleSuccess(data) {
  const response = validateData(data);
  let buttonDisabled = false;
  if (
    response.usernameError === ''
    && response.emailError === ''
    && response.passwordError === ''
    && response.cpasswordError === ''
  ) {
    buttonDisabled = false;
  } else {
    buttonDisabled = true;
  }
  return buttonDisabled;
}
