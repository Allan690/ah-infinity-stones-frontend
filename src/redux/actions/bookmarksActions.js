import { REQUEST_BOOKMARKS, BOOKMARKS_SUCCESS, BOOKMARKS_FAILURE } from './actionTypes';

export const requestBookmarks = () => ({
  type: REQUEST_BOOKMARKS,
});

export const bookmarksSuccess = data => ({
  type: BOOKMARKS_SUCCESS,
  data,
});

export const bookmarksFailure = error => ({
  type: BOOKMARKS_FAILURE,
  error,
});
