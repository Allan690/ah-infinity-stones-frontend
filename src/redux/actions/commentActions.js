import { GET_COMMENTS, POST_COMMENT, GET_TOPIC } from './actionTypes';

export const getComments = () => ({
  type: GET_COMMENTS,
});

export const postComment = postData => ({
  type: POST_COMMENT,
  payload: postData,
});

export const getTopic = () => ({
  type: GET_TOPIC,
});
