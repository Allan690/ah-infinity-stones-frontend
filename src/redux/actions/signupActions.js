import { SIGNUP_REQUEST, SIGNUP_SUCCESS, SIGNUP_FAILURE } from './actionTypes';

export const signupRequestAction = payload => ({
  type: SIGNUP_REQUEST,
  payload,
});

export const signupSuccessAction = payload => ({
  type: SIGNUP_SUCCESS,
  payload,
});

export const signupFailureAction = payload => ({
  type: SIGNUP_FAILURE,
  payload,
});
