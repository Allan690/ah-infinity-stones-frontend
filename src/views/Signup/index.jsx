import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import validator, {handleSuccess} from './validate';
import { successMessage} from '../../helpers/messages';
import { signupRequestAction } from '../../redux/actions/signupActions';
import SignupForm from '../../components/Forms/SignupForm'
import Loader from '../Loader';

class SignUp extends Component{
    state = {
      username: '',
      email: '',
      password:'',
      cpassword: ''
    };
    pageLoading = {
      loading: false,
    };
  handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value
      });
    }
  clearInputs = () => {
    this.setState({
      ...this.state,
      username: '',
      email: '',
      password: '',
      cpassword: ''
    });
    }
    handleSubmit = (e) => {
      e.preventDefault();
      let formValidate = validator(this.state);
      if(document.getElementById('serverMsgs').innerHTML === ""){
        this.pageLoading.loading = true;
      }
      else{
        this.pageLoading.loading = false;
      }
      if(formValidate.success !== ""){
      const { signupAction  } = this.props;
      signupAction (this.state);
    }
    };
    render(){
        let{error, success} = this.props;
        let messages = validator(this.state);
        if (success.Message) {
        this.clearInputs();
        successMessage(success.Message);
        return <Redirect to="/" />;
        }
        return (
            <div className="container" id="loginContainer">
                {this.pageLoading.loading === true ? <Loader /> : null}
                <SignupForm
                    handleChange = {this.handleChange}
                    handleSubmit = {this.handleSubmit}
                    messages = {messages}
                    error = {error}
                    handleSuccess = {handleSuccess(this.state)}
                />
            </div>
        )
    }
}
const mapStateToProps = state => ({
  error: state.signupReducer.error,
  success: state.signupReducer.success,
  payload: state.signupReducer.payload,
  signedUp: state.signupReducer.signedUp
});

const mapDispatchToProps = { signupAction: signupRequestAction};
export default connect(mapStateToProps, mapDispatchToProps)(SignUp);