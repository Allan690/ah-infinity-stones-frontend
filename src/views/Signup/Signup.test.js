import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Signup from '.';
import * as types from '../../redux/actions/actionTypes';
import * as actions from '../../redux/actions/signupActions';
import reducer from '../../redux/reducers/signupReducer';
import validateData, { handleSuccess } from './validate';
import responseData from '../../services/signupPayload';


describe('<Signup />', () => {
  let component;
  const initialState = {};
  const mockStore = configureStore();
  const store = mockStore(initialState);

  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        {' '}
        <Signup />
        {' '}
      </Provider>,
    );
  });
  it('render 1 Signup compoment', () => {
    expect(component).toHaveLength(1);
  });

  it('singup compoment should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('signup component contains a state', () => {
    expect(component.find('state'));
  });

  it('contains a user registration form', () => {
    expect(component.find('form'));
  });

  it('signup component contains a button', () => {
    expect(component.find('button'));
  });
});

describe('Action creators', () => {
  let payload = {};

  beforeEach(() => {
    payload = {
      user: {
        username: 'GuruMaster',
        email: 'gurumaster@gmail.com',
        password: 'Guru@2018',
      },
    };
  });

  it('should create an action to send signup request', () => {
    const expectedAction = {
      type: types.SIGNUP_REQUEST,
      payload,
    };
    expect(actions.signupRequestAction(payload)).toEqual(expectedAction);
  });

  it('should return an action for successful registration', () => {
    const expectedAction = {
      type: types.SIGNUP_SUCCESS,
      payload,
    };
    expect(actions.signupSuccessAction(payload)).toEqual(expectedAction);
  });

  it('should return an action for signup failure', () => {
    const expectedAction = {
      type: types.SIGNUP_FAILURE,
      payload,
    };
    expect(actions.signupFailureAction(payload)).toEqual(expectedAction);
  });
});

describe('Reducers', () => {
  let payload = {};

  beforeEach(() => {
    payload = {
      user: {
        username: 'Junior',
        email: 'junior@gmail.com',
        password: 'Junior@2019',
      },
    };
  });

  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(
      {
        error: {},
        payload: {},
        signedUp: false,
        success: {},
      },
    );
  });

  it('should handle SIGNUP_REQUEST', () => {
    expect(
      reducer(undefined, {
        type: types.SIGNUP_REQUEST,
        payload,
      }),
    ).toEqual(
      {
        error: {},
        payload,
        signedUp: false,
        success: {},
      },
    );
  });

  it('should handle SIGNUP_SUCCESS', () => {
    expect(
      reducer(undefined, {
        type: types.SIGNUP_SUCCESS,
        payload,
      }),
    ).toEqual(
      {
        error: {},
        payload: {},
        signedUp: true,
        success: {
          user: {
            email: 'junior@gmail.com',
            password: 'Junior@2019',
            username: 'Junior',
          },
        },
      },
    );
  });

  it('should handle SIGNUP_FAILURE', () => {
    expect(
      reducer(undefined, {
        type: types.SIGNUP_FAILURE,
        payload,
      }),
    ).toEqual(
      {
        error: payload,
        payload: {},
        signedUp: false,
        success: {},
      },
    );
  });
});

describe('validations', () => {
  it('can validate invalid username', () => {
    const name = {
      username: 'Paul',
      email: '',
      password: '',
      cpassword: '',
    };
    expect(validateData(name).usernameError).toEqual('Username must have at least six characters');
  });

  it('can validate invalid email', () => {
    const email = {
      username: 'Promaster',
      email: 'someone@gmailcom',
      password: '',
      cpassword: '',
    };
    expect(validateData(email).emailError).toEqual('Invalid email address');
  });

  it('can validate invalid password', () => {
    const password = {
      username: 'Gedion',
      email: 'gedion@gmail.com',
      password: 'Gedio',
      cpassword: '',
    };
    expect(validateData(password).passwordError).toEqual('Password must have at least six characters');
  });

  it('can validate invalid password', () => {
    const password = {
      username: 'Jackson',
      email: 'jackson@gmail.com',
      password: 'Jackson2018',
      cpassword: '',
    };
    expect(validateData(password).passwordError).toEqual('Enter strong password e.g p!Yd7YzU,S8}p}@');
  });

  it('can validate invalid cpassword', () => {
    const cpassword = {
      username: 'Someone',
      email: 'someone@gmail.com',
      password: 'Someone@2018',
      cpassword: 'Someone2018',
    };
    expect(validateData(cpassword).cpasswordError).toEqual('Password does not match');
  });

  it('can enable or disable sign up button', () => {
    const data = {
      username: 'Daniel',
      email: 'daniel@gmail.com',
      password: 'Daniel@2019',
      cpassword: 'Daniel@2019',
    };
    expect(handleSuccess(data).buttonDisabled).toBeFalsy();
  });
});

describe('Test Api call', () => {
  const data = {
    username: 'TestGuru',
    email: 'testguru@mail.com',
    password: 'TestGuru@2019',
    cpassword: 'TestGuru@2019',
  };
  const promise = responseData(data);

  it('A promise is returned from api call', () => {
    expect(promise).toBeTruthy();
  });

  it('Promise has the expected data', () => {
    promise.then((objects) => {
      expect(objects).toHaveLength(1);
    });
  });
});
