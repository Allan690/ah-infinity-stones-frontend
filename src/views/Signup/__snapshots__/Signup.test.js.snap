// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> singup compoment should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {},
    }
  }
>
   
  <Connect(SignUp) />
   
</ContextProvider>
`;
